src/mrinufft/_array_compat.py

```diff
@@ -0,0 +1,463 @@
+        if hasattr(_arg, "__array__"):
+            if is_cuda_array(_arg):
+                warnings.warn("data is on gpu, it will be moved to CPU.")
+            xp = get_array_module(_arg)
+            if xp.__name__ == "torch":
+                _arg = _arg.numpy(force=True)
+            elif xp.__name__ == "cupy":
+                _arg = cp.asnumpy(_arg)
+            elif "tensorflow" in xp.__name__:
+                _arg = _arg.numpy()
+        if isinstance(_arg, (tuple, list)):
+            _arg, _ = _to_numpy(*_arg)
+        kwargs[key] = _arg
+
+    return args, kwargs
+
+
+def _to_cupy(*args, device=None, **kwargs):
+    """Convert a sequence of arguments to cupy.
+
+    Non-arrays are ignored.
+
+    This avoid transfers between different devices (e.g., different GPUs).
+    """
+    # enforce mutable
+    args = list(args)
+
+    # convert positional arguments
+    for n in range(len(args)):
+        _arg = args[n]
+        if hasattr(_arg, "__array__"):
+            xp = get_array_module(_arg)
+            if xp.__name__ == "numpy":
+                with cp.cuda.Device(device):
+                    _arg = cp.asarray(_arg)
+            elif xp.__name__ == "torch":
+                if _arg.requires_grad:
+                    _arg = _arg.detach()
+                if _arg.is_cpu:
+                    with cp.cuda.Device(device):
+                        _arg = cp.asarray(_arg.numpy())
+                else:
+                    _arg = cp.from_dlpack(_arg)
+            elif "tensorflow" in xp.__name__:
+                if "CPU" in _arg.device:
+                    with cp.cuda.Device(device):
+                        _arg = cp.asarray(_arg.numpy())
+                else:
+                    _arg = cp.from_dlpack(tf.experimental.dlpack.to_dlpack(_arg))
+        if isinstance(_arg, (tuple, list)):
+            _arg, _ = _to_cupy(*_arg, device=device)
+
+        args[n] = _arg
+
+    # convert keyworded arguments
+    for key in kwargs.keys():
+        _arg = kwargs[key]
+        if hasattr(_arg, "__array__"):
+            xp = get_array_module(_arg)
+            if xp.__name__ == "numpy":
+                with cp.cuda.Device(device):
+                    _arg = cp.asarray(_arg)
+            elif xp.__name__ == "torch":
+                if _arg.requires_grad:
+                    _arg = _arg.detach()
+                if _arg.is_cpu:
+                    with cp.cuda.Device(device):
+                        _arg = cp.asarray(_arg.numpy())
+                else:
+                    _arg = cp.from_dlpack(_arg)
+            elif "tensorflow" in xp.__name__:
+                if "CPU" in _arg.device:
+                    with cp.cuda.Device(device):
+                        _arg = cp.asarray(_arg.numpy())
+                else:
+                    _arg = cp.from_dlpack(tf.experimental.dlpack.to_dlpack(_arg))
+        if isinstance(_arg, (tuple, list)):
+            _arg, _ = _to_cupy(*_arg, device=device)
+
+        kwargs[key] = _arg
```

Comment:
This is redundant with the above, lets refactor:
```suggestion
    # convert keyworded 
    process_kwargs_vals = _to_cupy(*kwargs.values())

    kwargs = {k: v for k,v in zip(kwargs.keys(), process_kwargs_vals}
```